### DISCUSSION TITLE
questions on fps for snapshot and refresh token subscribe help

### DISCUSSION THREAD

**Discussion by Leo7Deng** [dgreif/ring#1028] [Q&A]
I'm trying to iterate through gathering snapshots, my code looks like this:

const asyncLoop = async () => {
  const start = console.log(Date.now())

  for (let i = 0; i < 10; i++) {
    const data = await snapshot();
    console.log(Date.now() - start);
  }
}

asyncLoop();

I found that if I iterated through the loop too fast, it would just take a bunch of the image. The implementation I have above waits for the lyric to be saved before the next iteration but even at this point I found that it takes around 15 seconds for the next image to be shown. Is this a problem with my implementation or the ring service?

Also in my snapshot function I have a subscription to create a new refresh token but I found that this refresh token breaks after around 30-50 iterations, not sure why its such a big range. My refresh token does get updated every time but after the refresh token breaks, I have to manually generate a new one and put it back into the .env file. Has anyone had this problem before?

**Comment by dgreif** [comment]
The fps for snapshots is determined by Ring, and they error on the side of slowing down snapshots to prevent excessive battery/power usage on the cameras. You can _force_ the camera to take a snapshot, but that drains batteries quickly if you do it too frequently.

Regarding refresh token, you shouldn't need to manually trigger a token refresh. Ring is probably erroring because you are doing it too frequently. Instead, you should subscribe to the `onRefreshTokenUpdated` observable, like we do [here](https://github.com/dgreif/ring/blob/master/examples/example.ts#L21). This will automatically trigger any time the refresh token needed to be updated, and you can simply store the new value for the next time you create a `RingApi`
